### PR TITLE
Homework04a - Added possibility to search, filter books by title, author, isbn. 

### PR DESCRIPTION
**Pull request by noiseintheattic**

### CODE REVIEW

File: src/main/java/mate/academy/repository/book/IsbnSpecificationProvider.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import mate.academy.model.Book;
+import mate.academy.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class IsbnSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "isbn";
```
Comment (George-Khomiak) [inline src/main/java/mate/academy/repository/book/IsbnSpecificationProvider.java]: Let's create constants for all hardcoded strings

File: src/main/java/mate/academy/repository/book/AuthorSpecificationProvider.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+@Component
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "author";
+    }
+
+    public Specification<Book> getSpecification(String[] params) {
+        return (root, query, criteriaBuilder) -> root.get("author")
```
Comment (George-Khomiak) [inline src/main/java/mate/academy/repository/book/AuthorSpecificationProvider.java]: Let's have this author as a constant

File: src/main/java/mate/academy/repository/book/TitleSpecificationProvider.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import mate.academy.model.Book;
+import mate.academy.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "title";
```
Comment (George-Khomiak) [inline src/main/java/mate/academy/repository/book/TitleSpecificationProvider.java]: same here

File: src/main/java/mate/academy/repository/book/TitleSpecificationProvider.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    public Specification<Book> getSpecification(String[] params) {
+        return (root, query, criteriaBuilder) -> root.get("title")
+                .in(Arrays.stream(params).toArray());
+
+        /*
+        without lambda:
```
Comment (George-Khomiak) [inline src/main/java/mate/academy/repository/book/TitleSpecificationProvider.java]: Cool, but comments should be deleted

File: src/main/java/mate/academy/repository/BookSpecificationProviderManager.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import mate.academy.model.Book;
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
+@Component
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+
```
Comment (George-Khomiak) [inline src/main/java/mate/academy/repository/BookSpecificationProviderManager.java]: empty line

File: src/main/java/mate/academy/repository/BookSpecificationBuilder.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.RequiredArgsConstructor;
+import mate.academy.dto.BookSearchParametersDto;
+import mate.academy.model.Book;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
+@Component
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+
```
Comment (George-Khomiak) [inline src/main/java/mate/academy/repository/BookSpecificationBuilder.java]: Don't start class implementation with an empty line

**Review by George-Khomiak — APPROVED** [review]
Good job on handling git and great work overall. I left only minor comments. I know this task is hanging for a long time so I'll approve, but please apply these minor changes that I commented on in your next task :)

File: src/main/java/mate/academy/repository/book/IsbnSpecificationProvider.java (hunk 1 above)
Comment (noiseintheattic) [reply src/main/java/mate/academy/repository/book/IsbnSpecificationProvider.java]: ok, done

File: src/main/java/mate/academy/repository/book/AuthorSpecificationProvider.java (hunk 2 above)
Comment (noiseintheattic) [reply src/main/java/mate/academy/repository/book/AuthorSpecificationProvider.java]: ok, done

File: src/main/java/mate/academy/repository/book/TitleSpecificationProvider.java (hunk 3 above)
Comment (noiseintheattic) [reply src/main/java/mate/academy/repository/book/TitleSpecificationProvider.java]: ok, done

File: src/main/java/mate/academy/repository/book/TitleSpecificationProvider.java (hunk 4 above)
Comment (noiseintheattic) [reply src/main/java/mate/academy/repository/book/TitleSpecificationProvider.java]: ok, done

File: src/main/java/mate/academy/repository/BookSpecificationProviderManager.java (hunk 5 above)
Comment (noiseintheattic) [reply src/main/java/mate/academy/repository/BookSpecificationProviderManager.java]: ok, done

File: src/main/java/mate/academy/repository/BookSpecificationBuilder.java (hunk 6 above)
Comment (noiseintheattic) [reply src/main/java/mate/academy/repository/BookSpecificationBuilder.java]: ok, done

**Pull request merged**